README.md
# NAME

Apache2::Filter::Minifier::JavaScript - JavaScript minifying output filter for mod\_perl

# SYNOPSIS

```perl
<LocationMatch "\.js$">
    PerlOutputFilterHandler   Apache2::Filter::Minifier::JavaScript

    # if you need to supplement MIME-Type list
    PerlSetVar                JsMimeType  text/json

    # if you want to explicitly specify the minifier to use
    #PerlSetVar               JsMinifier  JavaScript::Minifier::XS
    #PerlSetVar               JsMinifier  JavaScript::Minifier
    #PerlSetVar               JsMinifier  MY::Minifier::function
</LocationMatch>
```

# DESCRIPTION

`Apache2::Filter::Minifier::JavaScript` is a Mod\_perl2 output filter which
minifies JavaScript using `JavaScript::Minifier` or
`JavaScript::Minifier::XS`.

Only JavaScript documents are minified, all others are passed through
unaltered.  `Apache2::Filter::Minifier::JavaScript` comes with a list of
several known acceptable MIME-Types for JavaScript documents, but you can
supplement that list yourself by setting the `JsMimeType` PerlVar
appropriately (use `PerlSetVar` for a single new MIME-Type, or `PerlAddVar`
when you want to add multiple MIME-Types).

Given a choice, using `JavaScript::Minifier::XS` is preferred over
`JavaScript::Minifier`, but we'll use whichever one you've got available.  If
you want to explicitly specify which minifier you want to use, set the
`JsMinifier` PerlVar to the name of the package/function that implements the
minifier.  Minification functions are expected to accept a single parameter
(the JavaScript to be minified) and to return the minified JavaScript on
completion.  If you specify a package name, we look for a `minify()` function
in that package.

## Caching

Minification does require additional CPU resources, and it is recommended that
you use some sort of cache in order to keep this to a minimum.

Being that you're already running Apache2, though, here's some examples of a
mod\_cache setup:

Disk Cache

```
# Cache root directory
CacheRoot /path/to/your/disk/cache
# Enable cache for "/js/" location
CacheEnable disk /js/
```

Memory Cache

```
# Cache size: 4 MBytes
MCacheSize 4096
# Min object size: 128 Bytes
MCacheMinObjectSize 128
# Max object size: 512 KBytes
MCacheMaxObjectSize 524288
# Enable cache for "/js/" location
CacheEnable mem /js/
```

# METHODS

- handler($filter)

    JavaScript minification output filter.

# AUTHOR

Graham TerMarsch (cpan@howlingfrog.com)

Many thanks to Geoffrey Young for writing `Apache::Clean`, from which several
things were lifted. :)

# COPYRIGHT

Copyright (C) 2007, Graham TerMarsch.  All Rights Reserved.

This is free software; you can redistribute it and/or modify it under the same
license as Perl itself.

# SEE ALSO

- [JavaScript::Minifier](https://metacpan.org/pod/JavaScript%3A%3AMinifier)
- [JavaScript::Minifier::XS](https://metacpan.org/pod/JavaScript%3A%3AMinifier%3A%3AXS)
- [Apache2::Filter::Minifier::CSS](https://metacpan.org/pod/Apache2%3A%3AFilter%3A%3AMinifier%3A%3ACSS)
- [Apache::Clean](https://metacpan.org/pod/Apache%3A%3AClean)


Changes
Revision history for Perl extension Apache2::Filter::Minifier::JavaScript.

{{$NEXT}}

1.07      2021-01-17 11:52:20-08:00 America/Vancouver
    - Switch to Dist::Zilla for release management
    - POD cleanups; spelling
    - Namespace cleaned
    - Add a GitHub actions CI workflow
      - but unfortunately, only for Perl-5.20 or higher

1.06    Fri Nov  7 17:05 PST 2014
    - Update tests to accommodate Apache-2.4.x series

1.05    Fri Apr 23 22:25 PDT 2010
    - Switch to Git

1.04    Mon May  5 15:17 PDT 2008
    - no changes since 1.04_02, but is a non-devel release

1.04_02 Wed Oct 17 16:01 PDT 2007
    - decline processing if unable to determine Content-Type of document
    - lower Perl requirement to 5.6.0

1.04_01 Tue Oct 16 00:59 PDT 2007
    - added timing info to the debug logs
    - use JavaScript::Minifier::XS if available; its -MUCH- faster
    - new "JsMinifier" PerlVar, to let you specify which minifier package or
      function to use (otherwise we try J:M:XS first, falling back to J:M)

1.03    Fri Sep 28 11:09 PDT 2007
    - release; verified that changes in 1.03_01 for configuration requirements
      worked as intended when doing CPANPLUS test runs

1.03_01 Thu Sep 27 13:27 PDT 2007
    - rewrote Build.PL/Makefile.PL so they now check if you've got Apache::Test
      installed -at-configuration-time-.  If not, the build is aborted.  One
      day, CPAN(PLUS) et al. will support "configure_requires", but until then
      we've got to check for this ourselves.
    - fixed Content-Type filtering so that it now properly handles HTTP headers
      that have additional attributes (e.g. "charset=...")

1.02    Wed Sep 26  1:25 PDT 2007
    - added Makefile.PL, so CPANPLUS testers stop failing tests because they
      don't have Module::Build installed

1.01    Fri Sep 21 13:31 PDT 2007
    - added explicit dependency on MP2
    - updated test requirements to "need_*" instead of "have_*"

1.00    Thu Sep 20 12:45 PDT 2007
    - initial version


Build.PL
use 5.006;
use strict;
use warnings;
use Module::Build;

###############################################################################
# Check for presence of Apache::Test.
#
# If its not installed, we're going to abort the build entirely as we'll be
# unable to run our test suites (which need to be set up -NOW- and not when you
# run "./Build").  The joys of "configure requirements"...
#
# Fortunately, Apache::Test comes -with- mod_perl2, so anyone who actually
# wants to use this module probably already has it installed.  If not, we abort
# -before- we write out "Build", so that CPAN(PLUS) won't file CPAN Tester
# "failure" reports; even if they were to install Apache::Test as a dependency
# we'd still fail our tests... its a -configuration-requirement-.
my $HAVE_APACHE_TEST = eval {
        require Apache::TestMB
        };
unless ($HAVE_APACHE_TEST) {
    # show warning
    warn q{
###
### This module requires that you have Apache::Test installed -before-
### you run "perl Build.PL", so that test suites can be properly
### configured.
###
### Please go install Apache::Test and re-run "perl Build.PL".
###
};
    # abort, before writing Build script
    exit 0;
}

###############################################################################
# Packages we require at -configuration- time
our %configure_requires = (
    'Apache::Test'  => 1.12,
    );

###############################################################################
# Write Build script
my $build_pkg = $HAVE_APACHE_TEST
                    ? 'Apache::TestMB'
                    : 'Module::Build';
$build_pkg->new(
    'module_name'       => 'Apache2::Filter::Minifier::JavaScript',
    'license'           => 'perl',
    'dist_author'       => 'Graham TerMarsch (cpan@howlingfrog.com)',
    'requires'          => {
        'JavaScript::Minifier'  => 0,
        'mod_perl2'             => 2.0,
        'Time::HiRes'           => 0,
        'perl'                  => '5.6.0',
        },
    'recommends'        => {
        'JavaScript::Minifier::XS' => 0,
        },
    'test_requires' => {
        'File::Slurp' => 0,
    },
    # add configuration requirements both to build_requires and to META.yml,
    # until M::B supports "configure_requires" directly.
    'build_requires'    => { %configure_requires },
    'meta_add'          => {
        'configure_requires' => { %configure_requires },
        },
    )->create_build_script();


lib/Apache2/Filter/Minifier/JavaScript.pm
package Apache2::Filter::Minifier::JavaScript;

###############################################################################
# Required inclusions.
use strict;
use warnings;
use Apache2::Filter qw();           # $f
use Apache2::RequestRec qw();       # $r
use Apache2::RequestUtil qw();      # $r->dir_config
use Apache2::Log qw();              # $log->*()
use APR::Table qw();                # dir_config->get() and headers_out->unset()
use Apache2::Const -compile => qw(OK DECLINED);
use Time::HiRes qw(gettimeofday tv_interval);

###############################################################################
# Load up the JS minifier modules.
use JavaScript::Minifier qw(minify);
eval { require JavaScript::Minifier::XS; };

###############################################################################
# Keep our namespace clean.
use namespace::clean;

###############################################################################
# Version number.
our $VERSION = '1.08';

###############################################################################
# MIME-Types we're willing to minify.
my %mime_types = (
    'text/javascript'           => 1,
    'text/ecmascript'           => 1,
    'application/javascript'    => 1,
    'application/ecmascript'    => 1,
    'application/x-javascript'  => 1,
    );

###############################################################################
# Subroutine:   handler($filter)
###############################################################################
# JavaScript minification output filter.
sub handler {
    my $f = shift;
    my $r = $f->r;
    my $log = $r->log;

    # assemble list of acceptable MIME-Types
    my %types = (
        %mime_types,
        map { $_=>1 } $r->dir_config->get('JsMimeType'),
        );

    # determine Content-Type of document
    my ($ctype) = ($r->content_type =~ /^(.+?)(?:;.*)?$/);
    unless ($ctype) {
        $log->info( "unable to determine content type; skipping : URL ", $r->uri );
        return Apache2::Const::DECLINED;
    }

    # only process JS documents
    unless (exists $types{$ctype}) {
        $log->info( "skipping request to ", $r->uri, " (not a JS document)" );
        return Apache2::Const::DECLINED;
    }

    # figure out which minifier module/function we're supposed to be using;
    # either an explicit minifier function/package, or our list of acceptable
    # minifiers
    my $minifier;
    my @possible = $r->dir_config->get('JsMinifier') || (
        'JavaScript::Minifier::XS',
        'JavaScript::Minifier',
        );
    foreach my $maybe (@possible) {
        no strict 'refs';
        # explicit function name
        if (defined &{"$maybe"}) {
            $minifier = sub { $maybe->(shift) };
            last;
        }
        # package name; look for "minify()" function
        if (defined &{"${maybe}::minify"}) {
            my $func = \&{"${maybe}::minify"};
            $minifier = ($maybe eq 'JavaScript::Minifier')
                            ? sub { $func->(input=>shift) }
                            : sub { $func->(shift) };
            last;
        }
    }
    unless ($minifier) {
        $log->info( "no JavaScript minifier available; declining" );
        return Apache2::Const::DECLINED;
    }

    # gather up entire document
    my $ctx = $f->ctx;
    while ($f->read(my $buffer, 4096)) {
        $ctx .= $buffer;
    }

    # unless we're at the end, store the JS for our next invocation
    unless ($f->seen_eos) {
        $f->ctx( $ctx );
        return Apache2::Const::OK;
    }

    # if we've got JS to minify, minify it
    if ($ctx) {
        my $t_st = [gettimeofday()];
        my $min  = eval { $minifier->($ctx) };
        if ($@) {
            # minification failed; log error and send original JS
            $log->error( "error minifying: $@" );
            $f->print( $ctx );
        }
        else {
            # minification ok; log results and send minified JS
            my $t_dif = tv_interval($t_st);
            my $l_min = length($min);
            my $l_js  = length($ctx);
            $log->debug( "JS minified $l_js to $l_min : t:$t_dif : URL ", $r->uri );
            $r->headers_out->unset( 'Content-Length' );
            $f->print( $min );
        }
    }

    return Apache2::Const::OK;
}

1;

=for stopwords PerlVar minified minifier minification minifies minifying

=head1 NAME

Apache2::Filter::Minifier::JavaScript - JavaScript minifying output filter for mod_perl

=head1 SYNOPSIS

  <LocationMatch "\.js$">
      PerlOutputFilterHandler   Apache2::Filter::Minifier::JavaScript

      # if you need to supplement MIME-Type list
      PerlSetVar                JsMimeType  text/json

      # if you want to explicitly specify the minifier to use
      #PerlSetVar               JsMinifier  JavaScript::Minifier::XS
      #PerlSetVar               JsMinifier  JavaScript::Minifier
      #PerlSetVar               JsMinifier  MY::Minifier::function
  </LocationMatch>

=head1 DESCRIPTION

C<Apache2::Filter::Minifier::JavaScript> is a Mod_perl2 output filter which
minifies JavaScript using C<JavaScript::Minifier> or
C<JavaScript::Minifier::XS>.

Only JavaScript documents are minified, all others are passed through
unaltered.  C<Apache2::Filter::Minifier::JavaScript> comes with a list of
several known acceptable MIME-Types for JavaScript documents, but you can
supplement that list yourself by setting the C<JsMimeType> PerlVar
appropriately (use C<PerlSetVar> for a single new MIME-Type, or C<PerlAddVar>
when you want to add multiple MIME-Types).

Given a choice, using C<JavaScript::Minifier::XS> is preferred over
C<JavaScript::Minifier>, but we'll use whichever one you've got available.  If
you want to explicitly specify which minifier you want to use, set the
C<JsMinifier> PerlVar to the name of the package/function that implements the
minifier.  Minification functions are expected to accept a single parameter
(the JavaScript to be minified) and to return the minified JavaScript on
completion.  If you specify a package name, we look for a C<minify()> function
in that package.

=head2 Caching

Minification does require additional CPU resources, and it is recommended that
you use some sort of cache in order to keep this to a minimum.

Being that you're already running Apache2, though, here's some examples of a
mod_cache setup:

Disk Cache

  # Cache root directory
  CacheRoot /path/to/your/disk/cache
  # Enable cache for "/js/" location
  CacheEnable disk /js/

Memory Cache

  # Cache size: 4 MBytes
  MCacheSize 4096
  # Min object size: 128 Bytes
  MCacheMinObjectSize 128
  # Max object size: 512 KBytes
  MCacheMaxObjectSize 524288
  # Enable cache for "/js/" location
  CacheEnable mem /js/

=head1 METHODS

=over

=item handler($filter)

JavaScript minification output filter.

=back

=head1 AUTHOR

Graham TerMarsch (cpan@howlingfrog.com)

Many thanks to Geoffrey Young for writing C<Apache::Clean>, from which several
things were lifted. :)

=head1 COPYRIGHT

Copyright (C) 2007, Graham TerMarsch.  All Rights Reserved.

This is free software; you can redistribute it and/or modify it under the same
license as Perl itself.

=head1 SEE ALSO

=over

=item L<JavaScript::Minifier>

=item L<JavaScript::Minifier::XS>

=item L<Apache2::Filter::Minifier::CSS>

=item L<Apache::Clean>

=back

=cut
